stream/harfanglab/src/harfanglab.py

```diff
@@ -152,422 +168,526 @@ def _process_message(self, msg):
-            indicator_name = data["name"]
-            get_indicator = self._query(
-                "get", f"/{uri}/?search={indicator_name}&source_id={source_list_id}"
-            )
-            indicator_matched = self.find_data_name_match(
-                get_indicator["results"], indicator_name
-            )
-
+    def process_update_indicator(
+        self, data, entity, uri, pattern_type, indicator_matched
+    ):
         if indicator_matched is None:
             # UPSERT
             if data["pattern_type"] == "yara":
                 self.create_indicator(data, "yara", "YaraFile", self.yara_list_id)
             elif data["pattern_type"] == "sigma":
                 self.create_indicator(data, "sigma", "SigmaRule", self.sigma_list_id)
             elif data["pattern_type"] == "stix":
-                entity = self.helper.api.indicator.read(
-                    id=OpenCTIConnectorHelper.get_attribute_in_extension("id", data)
-                )
-                self.create_observable(entity, "stix", "IOCRule")
+                self.create_indicator(data, "stix", "IOCRule", self.stix_list_id)
             else:
-                raise ValueError("Unsupported Pattern")
+                return self.helper.log_error("Unsupported Pattern")
             return
-
         else:
-            data["content"] = data["pattern"]
-            payload = self.pattern_payload(data)
             indicator_id = indicator_matched["id"]
-            response = self._query("put", f"/{uri}/{indicator_id}/", payload)
+            if pattern_type == "stix":
+                indicator = self.build_stix_indicator_object(
+                    indicator_matched, entity, indicator_matched["enabled"], True
+                )
+                indicator["source"] = indicator["source_id"]
+            else:
+                indicator = self.build_yara_sigma_indicator_object(
+                    data, entity, indicator_matched["enabled"]
+                )
+
+            response = self._query("put", f"/{uri}/{indicator_id}/", indicator)
+            if response is None:
+                return self.helper.log_error(
+                    f"[UPDATE] Failure {pattern_type} indicator updated = {indicator_id}"
+                )
+            else:
+                return self.helper.log_info(
+                    f"[UPDATE] Successful {pattern_type} indicator updated = {indicator_id}"
+                )
+
+    def process_update_observable(
+        self, data, entity, uri, pattern_type, observable_matched
+    ):
+        if observable_matched is None:
+            msg_log = f"[UPDATE] The searched for {pattern_type} observable does not exist in HarfangLab"
+            return self.helper.log_error(msg_log)
+        else:
+            observable_id = observable_matched["id"]
+            observable_matched["description"] = data.get(
+                "description", "No description"
+            )
+            observable = self.build_stix_indicator_object(
```

Comment:
I don't understand why we are calling `build_stix_indicator_object` in `process_update_observable`, can you help me ?

Reply:
Indeed thank you for the feedback